util/rowDecoder/decoder.go

```diff
@@ -134,3 +136,74 @@ func (rd *RowDecoder) DecodeAndEvalRowWithMap(ctx sessionctx.Context, handle int
 	}
 	return row, nil
 }
+
+// SubstituteGenColsInDecodeColMap substitutes generated columns in every expression in decodeColMap
+// with non-generated one by looking up decodeColMap.
+func SubstituteGenColsInDecodeColMap(decodeColMap map[int64]Column) {
+	// Sort columns by ID in ascending order.
+	var orderedCols []int64
+	for colID := range decodeColMap {
+		orderedCols = append(orderedCols, colID)
+	}
+	sort.Slice(orderedCols, func(i, j int) bool { return orderedCols[i] < orderedCols[j] })
+
+	// Iterate over decodeColMap, substitution only happens once because
+	// the column with smaller columnID can not refer to those with larger.
+	for _, colID := range orderedCols {
+		decCol := decodeColMap[colID]
+		if decCol.GenExpr != nil {
+			decodeColMap[colID] = Column{
+				Col:     decCol.Col,
+				GenExpr: substituteGeneratedColumn(decCol.GenExpr, decodeColMap),
+			}
+		} else {
+			decodeColMap[colID] = Column{
+				Col: decCol.Col,
+			}
+		}
+	}
+}
+
+// substituteGeneratedColumn substitutes generated columns in an expression with non-generated one by looking up decodeColMap.
+func substituteGeneratedColumn(expr expression.Expression, decodeColMap map[int64]Column) expression.Expression {
+	switch v := expr.(type) {
+	case *expression.Column:
+		if c, ok := decodeColMap[v.ID]; ok && c.GenExpr != nil {
+			return c.GenExpr
+		}
+		return v
+	case *expression.ScalarFunction:
+		if v.FuncName.L == ast.Cast {
```

Comment:
Why do we handle `CAST` in this special way? I don't get the point from `expression.ColumnSubstitute`, neither from here.

Reply:
Actually, I am not sure. The removal of them did NOT fail the integration test.